libflux/src/semantic/types.rs

```diff
@@ -959,8 +959,71 @@ impl MaxTvar for Function {
+        //  3. If a pipe arg was used then the other type must specify a pipe arg too.
+        // This is done at the beginning both to avoid starting unifying when there is no need to,
+        // and in order to error using f and g, otherwise, the borrow checker would not make us
+        // error while we have borrowed some of their internal pieces (while iterating for example).
+        if f.req
+            .keys()
+            .any(|k| !g.req.contains_key(k) && !g.opt.contains_key(k))
+            || g.req
+                .keys()
+                .any(|k| !f.req.contains_key(k) && !f.opt.contains_key(k))
+            || match (&f.pipe, &g.pipe) {
+                (Some(_), None) | (None, Some(_)) => true,
+                _ => false,
+            }
+        {
+            return Err(Error::cannot_unify(&f, &g));
+        }
+        // Given two function types f and g, the process for unifying their arguments is as follows:
+        // 1. If a required arg of f is not present in the arguments of g, fail unification (checked above),
+        //    otherwise unify both argument types.
+        let sub: Substitution =
+            f.req
+                .into_iter()
+                .try_fold(Substitution::empty(), |sub, (arg_name, f_arg_type)| {
+                    if let Some(g_arg_type) = g.req.remove(arg_name.as_str()) {
+                        apply_then_unify(f_arg_type, g_arg_type, sub, cons, fresh)
+                    } else if let Some(g_arg_type) = g.opt.remove(arg_name.as_str()) {
+                        apply_then_unify(f_arg_type, g_arg_type, sub, cons, fresh)
+                    } else {
+                        panic!("this should never happen, checked above");
+                    }
+                })?;
+        // 2. If an optional arg of f is not present in the arguments of g, continue,
+        //    otherwise unify both argument types (repeat for g).
+        let mut sub = f
+            .opt
+            .into_iter()
+            .try_fold(sub, |sub, (arg_name, f_arg_type)| {
+                if let Some(g_arg_type) = g.req.remove(arg_name.as_str()) {
+                    apply_then_unify(f_arg_type, g_arg_type, sub, cons, fresh)
+                } else if let Some(g_arg_type) = g.opt.remove(arg_name.as_str()) {
+                    apply_then_unify(f_arg_type, g_arg_type, sub, cons, fresh)
+                } else {
+                    Ok(sub)
+                }
+            })?;
+        // 3. Lastly unify pipe args. Note that pipe arguments are optional.
+        //    However if a pipe arg was used in a calling context, i.e it's an un-named pipe arg,
+        //    then the other type must specify a pipe arg too, otherwise unification fails (checked above).
+        if let (Some(Property { k: _, v: pf }), Some(Property { k: _, v: pg })) = (f.pipe, g.pipe) {
+            sub = apply_then_unify(pf, pg, sub, cons, fresh)?;
+        }
+        Ok(sub)
```

Comment:
Don't forget to unify the return types. I would probably just unify those right off the bat since they're the simplest.